### ISSUE TITLE
Panic with 1.19.5+k3s1 on rpi-4 server

### ISSUE THREAD

**Issue by narqo** [k3s-io/k3s#2704]
**Environmental Info:**
K3s Version: 
```
pi@pi-41:~ $ k3s -v
k3s version v1.19.5+k3s1 (b11612e2)
```

Node(s) CPU architecture, OS, and Version: 
```
pi@pi-41:~ $ uname -a
Linux pi-41 5.4.79-v7l+ #1373 SMP Mon Nov 23 13:27:40 GMT 2020 armv7l GNU/Linux
```

Cluster Configuration:
4 nodes cluster running on rpi-4s (2GB): 1 server + 3 workers

**Describe the bug:**
After updating the cluster from v1.19.4+k3s2 to v1.19.5+k3s1, the server starts panicking with `unexpected fault address` on server's startup.

From the logs (see below) it looks like the crash happens after the server reaches "CronJob Manager". In fact, the only workload I run on the server for the last couple months was [this cronjob](https://github.com/narqo/github-weather/blob/v3.1.1/deployments/cronjob.yaml).

**Steps To Reproduce:**

- Installed K3s: 

**Expected behavior:**

**Actual behavior:**

**Additional context / logs:**

<details>
<summary>An extract from the stack trace</summary>

```
Dec 14 23:30:02 pi-41 k3s[525]: I1214 23:30:02.891889     525 shared_informer.go:240] Waiting for caches to sync for disruption
Dec 14 23:30:02 pi-41 k3s[525]: I1214 23:30:02.984144     525 controllermanager.go:549] Started "cronjob"
Dec 14 23:30:02 pi-41 k3s[525]: I1214 23:30:02.984415     525 cronjob_controller.go:96] Starting CronJob Manager
Dec 14 23:30:03 pi-41 k3s[525]: unexpected fault address 0x123ee0d0
Dec 14 23:30:03 pi-41 k3s[525]: fatal error: fault
Dec 14 23:30:03 pi-41 k3s[525]: [signal SIGSEGV: segmentation violation code=0x2 addr=0x123ee0d0 pc=0x123ee0d0]
Dec 14 23:30:03 pi-41 k3s[525]: goroutine 43396 [running]:
Dec 14 23:30:03 pi-41 k3s[525]: runtime.throw(0x3372b14, 0x5)
Dec 14 23:30:03 pi-41 k3s[525]:         /usr/local/go/src/runtime/panic.go:1116 +0x5c fp=0x11b60b90 sp=0x11b60b7c pc=0x46154
Dec 14 23:30:03 pi-41 k3s[525]: runtime.sigpanic()
Dec 14 23:30:03 pi-41 k3s[525]:         /usr/local/go/src/runtime/signal_unix.go:749 +0x310 fp=0x11b60ba8 sp=0x11b60b90 pc=0x5d688
Dec 14 23:30:03 pi-41 k3s[525]: runtime: unexpected return pc for github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus called from 0x0
Dec 14 23:30:03 pi-41 k3s[525]: stack: frame={sp:0x11b60bac, fp:0x11b60c5c} stack=[0x11b5e000,0x11b62000)
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b2c:  00047eb0 <runtime.gwrite+52>  057e8698  0336ead4  00000001
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b3c:  00000001  00000000  00048748 <runtime.printstring+96>  00000002
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b4c:  0336ead4  00000001  00000001  00000000
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b5c:  00000000  000462f4 <runtime.fatalthrow+68>  00046154 <runtime.throw+92>  11b60b6c
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b6c:  000748fc <runtime.fatalthrow.func1+0>  0e767ea0  00046154 <runtime.throw+92>  11b60b7c
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b7c:  0005d688 <runtime.sigpanic+784>  11b60b84  00074890 <runtime.throw.func1+0>  03372b14
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b8c:  00000005  0207c614 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus+88>  03372b14  00000005
Dec 14 23:30:03 pi-41 k3s[525]: 11b60b9c:  00000001  123ee0d0  0e767ea0  0207c614 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus+88>
Dec 14 23:30:03 pi-41 k3s[525]: 11b60bac: <00000000  00000001  a6915110  00000000
Dec 14 23:30:03 pi-41 k3s[525]: 11b60bbc:  00000064  123ee0d0  00085684 <sync.(*Pool).pin+48>  00000065
Dec 14 23:30:03 pi-41 k3s[525]: 11b60bcc:  00000016  b6ee93ec  00085684 <sync.(*Pool).pin+48>  0008524c <sync.(*Pool).Put+40>
Dec 14 23:30:03 pi-41 k3s[525]: 11b60bdc:  05577fe4  00000004  00000700  00000001
Dec 14 23:30:03 pi-41 k3s[525]: 11b60bec:  00117e10 <fmt.(*pp).free+160>  00022e08 <runtime.mapassign_faststr+72>  18d901b0  5b731b39
Dec 14 23:30:03 pi-41 k3s[525]: 11b60bfc:  00000024  0f19b06c  02079030 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.syncOne+412>  11b60c40
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c0c:  5b731b39  0f19b06c  15142300  0000000f
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c1c:  0d08ac80  00000016  0f19b06c  00000016
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c2c:  0d08ac80  11b611bc  0207ce60 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.(*Controller).syncAll.func4+208>  11b61210
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c3c:  00000000  00000000  00000000  139ea950
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c4c:  0000000e  00000000  00000000  139ea960
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c5c: >00000007  09fedbc0  0000003e  18cbfd70
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c6c:  00000024  139ea967  00000007  00000000
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c7c:  00000000  00000000  00000000  d72a7c3f
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c8c:  0000000e  057d12a8  00000000  00000000
Dec 14 23:30:03 pi-41 k3s[525]: 11b60c9c:  00000000  0c4a2080  00000000  00000000
Dec 14 23:30:03 pi-41 k3s[525]: 11b60cac:  00000000  00000000  00000000  00000000
Dec 14 23:30:03 pi-41 k3s[525]: 11b60cbc:  00000000  00000000  130f8550  00000002
Dec 14 23:30:03 pi-41 k3s[525]: 11b60ccc:  00000002  139ea9d0  0000000c  00000000
Dec 14 23:30:03 pi-41 k3s[525]: github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus(0x9fedbc0, 0x3e, 0x18cbfd70, 0x24)
Dec 14 23:30:03 pi-41 k3s[525]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob/utils.go:175 +0x58 fp=0x11b60c5c sp=0x11b60bac pc=0x207c614
Dec 14 23:30:03 pi-41 k3s[525]: created by github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.(*Controller).Run
Dec 14 23:30:03 pi-41 k3s[525]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob/cronjob_controller.go:98 +0xf8
Dec 14 23:30:03 pi-41 k3s[525]: goroutine 1 [chan receive]:
Dec 14 23:30:03 pi-41 k3s[525]: github.com/rancher/k3s/pkg/agent.run(0x3a58288, 0x7766340, 0x139e49a0, 0x6a, 0x0, 0x0, 0x0, 0x0, 0xcfaf320, 0x16, ...)
Dec 14 23:30:03 pi-41 k3s[525]:         /go/src/github.com/rancher/k3s/pkg/agent/run.go:119 +0x1c0
Dec 14 23:30:03 pi-41 k3s[525]: github.com/rancher/k3s/pkg/agent.Run(0x3a58288, 0x7766340, 0x139e49a0, 0x6a, 0x0, 0x0, 0x0, 0x0, 0xcfaf320, 0x16, ...)
Dec 14 23:30:03 pi-41 k3s[525]:         /go/src/github.com/rancher/k3s/pkg/agent/run.go:170 +0x2f4
Dec 14 23:30:03 pi-41 k3s[525]: github.com/rancher/k3s/pkg/cli/server.run(0x774c370, 0x57d1ed0, 0x48, 0x9f0f34)
Dec 14 23:30:03 pi-41 k3s[525]:         /go/src/github.com/rancher/k3s/pkg/cli/server/server.go:283 +0x12c0
Dec 14 23:30:03 pi-41 k3s[525]: github.com/rancher/k3s/pkg/cli/server.Run(0x774c370, 0x0, 0x774c370)
```

</details>

Full logs [pi-41-k3s-daemon.1.log.gz](https://github.com/k3s-io/k3s/files/5692033/pi-41-k3s-daemon.1.log.gz)

**Comment by brandond** [comment]
This is a 2gb Pi4b? What OS and release are you running?

**Comment by narqo** [comment]
On server node (Pi4b, 2GB), I'm running Raspbian Linux (32-bit)

```
pi@pi-41:~ $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
NAME="Raspbian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
```

**Comment by brandond** [comment]
I'm unable to reproduce this on my Pi4b running Ubuntu. Are you up-to-date with OS and kernel updates?
```
root@pi02:~# uname -a
Linux pi02.lan.khaus 5.8.0-1008-raspi #11-Ubuntu SMP PREEMPT Thu Nov 12 15:56:06 UTC 2020 armv7l armv7l armv7l GNU/Linux
root@pi02:~# kubectl version -o yaml
clientVersion:
  buildDate: "2020-12-11T17:29:41Z"
  compiler: gc
  gitCommit: b11612e2744f39f01bfd208ff97315930c483667
  gitTreeState: clean
  gitVersion: v1.19.5+k3s1
  goVersion: go1.15.5
  major: "1"
  minor: "19"
  platform: linux/arm
serverVersion:
  buildDate: "2020-12-11T17:29:41Z"
  compiler: gc
  gitCommit: b11612e2744f39f01bfd208ff97315930c483667
  gitTreeState: clean
  gitVersion: v1.19.5+k3s1
  goVersion: go1.15.5
  major: "1"
  minor: "19"
  platform: linux/arm
```

**Comment by narqo** [comment]
I updated the Pi to the latest kernel but the problem still presents

```
pi@pi-41:~ $ uname -a
Linux pi-41 5.10.0-v7l+ #1382 SMP Tue Dec 15 18:23:34 GMT 2020 armv7l GNU/Linux

pi@pi-41:~ $ sudo ./dl/k3s-v1.19.5+k3s1 kubectl version -o yaml
clientVersion:
  buildDate: "2020-12-11T17:29:41Z"
  compiler: gc
  gitCommit: b11612e2744f39f01bfd208ff97315930c483667
  gitTreeState: clean
  gitVersion: v1.19.5+k3s1
  goVersion: go1.15.5
  major: "1"
  minor: "19"
  platform: linux/arm

The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
```

I can reproduce the problem by running k3s binary directly (attaching the full crash log [k3s-v1.19.5-server.log.gz](https://github.com/k3s-io/k3s/files/5703410/k3s-v1.19.5-server.log.gz))

```
pi@pi-41:~ $ sudo ./dl/k3s-v1.19.5+k3s1 server 2>&1 | tee k3s-v1.19.5-server.log

time="2020-12-16T15:11:36.675594099+01:00" level=info msg="Starting k3s v1.19.5+k3s1 (b11612e2)"
time="2020-12-16T15:11:36.684918636+01:00" level=info msg="Cluster bootstrap already complete"
time="2020-12-16T15:11:36.759955180+01:00" level=info msg="Configuring sqlite3 database connection pooling: maxIdleConns=2, maxOpenConns=0, connMaxLifetime=0s"
···
unexpected fault address 0x1049db40
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x2 addr=0x1049db40 pc=0x1049db40]

goroutine 45423 [running]:
runtime.throw(0x3372b14, 0x5)
        /usr/local/go/src/runtime/panic.go:1116 +0x5c fp=0xbd16b90 sp=0xbd16b7c pc=0x46154
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:749 +0x310 fp=0xbd16ba8 sp=0xbd16b90 pc=0x5d688
runtime: unexpected return pc for github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus called from 0x0
···
```

With `k3s-v1.19.4+k3s2` everything still works fine.

**Comment by narqo** [comment]
A quick update, I'm testing v1.20.0+k3s.1 on the server node of the same Pi cluster, it seems working

```
pi@pi-41:~ $ sudo kubectl get node -o wide
NAME    STATUS   ROLES                  AGE   VERSION            INTERNAL-IP     EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME
pi-44   Ready    <none>                 49d   v1.19.5+k3s1       192.168.20.44   <none>        Ubuntu 20.04.1 LTS               5.4.0-1023-raspi   containerd://1.4.3-k3s1
pi-42   Ready    <none>                 49d   v1.19.5+k3s1       192.168.20.42   <none>        Raspbian GNU/Linux 10 (buster)   5.4.79-v7l+        containerd://1.4.3-k3s1
pi-41   Ready    control-plane,master   49d   v1.20.0-rc3+k3s1   192.168.20.41   <none>        Raspbian GNU/Linux 10 (buster)   5.10.0-v7l+        containerd://1.4.3-k3s1
pi-43   Ready    <none>                 18d   v1.19.5+k3s1       192.168.20.43   <none>        Debian GNU/Linux 10 (buster)     5.4.79-v8+         containerd://1.4.3-k3s1
```

**Comment by ViBiOh** [comment]
I had the same issue with a rpi4 2GB RAM with raspbian (32 bits) up-to-date. I have a master and two workers. I always update the `master` first, and if everything is right, the workers then.

My configuration for the master

```
k3s server --disable-agent --data-dir "/opt/k3s/data" --disable traefik --disable servicelb --disable metrics-server --kube-apiserver-arg enable-admission-plugins=PodSecurityPolicy,NodeRestriction --kube-apiserver-arg enable-aggregator-routing=true
```

When bumping from `v1.19.4+k3s2` to `v1.19.5+k3s*` (both s1 and s2 fail), I got panic that start from that moment.

```
Dec 18 18:12:50 walter k3s[19344]: I1218 18:12:50.272136   19344 controllermanager.go:549] Started "job"
Dec 18 18:12:50 walter k3s[19344]: I1218 18:12:50.272764   19344 job_controller.go:148] Starting job controller
Dec 18 18:12:50 walter k3s[19344]: I1218 18:12:50.272812   19344 shared_informer.go:240] Waiting for caches to sync for job
Dec 18 18:12:50 walter k3s[19344]: unexpected fault address 0x1225ccd0
Dec 18 18:12:50 walter k3s[19344]: fatal error: fault
Dec 18 18:12:50 walter k3s[19344]: [signal SIGSEGV: segmentation violation code=0x2 addr=0x1225ccd0 pc=0x1225ccd0]
Dec 18 18:12:50 walter k3s[19344]: goroutine 24100 [running]:
Dec 18 18:12:50 walter k3s[19344]: runtime.throw(0x3372b14, 0x5)
Dec 18 18:12:50 walter k3s[19344]:         /usr/local/go/src/runtime/panic.go:1116 +0x5c fp=0x1b072b90 sp=0x1b072b7c pc=0x46154
Dec 18 18:12:50 walter k3s[19344]: runtime.sigpanic()
Dec 18 18:12:50 walter k3s[19344]:         /usr/local/go/src/runtime/signal_unix.go:749 +0x310 fp=0x1b072ba8 sp=0x1b072b90 pc=0x5d688
Dec 18 18:12:50 walter k3s[19344]: runtime: unexpected return pc for github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus called from 0x0
Dec 18 18:12:50 walter k3s[19344]: stack: frame={sp:0x1b072bac, fp:0x1b072c5c} stack=[0x1b070000,0x1b074000)
Dec 18 18:12:50 walter k3s[19344]: 1b072b2c:  00047eb0 <runtime.gwrite+52>  057e860e  0336ead4  00000001
Dec 18 18:12:50 walter k3s[19344]: 1b072b3c:  00000001  00000000  00048748 <runtime.printstring+96>  00000002
Dec 18 18:12:50 walter k3s[19344]: 1b072b4c:  0336ead4  00000001  00000001  00000000
Dec 18 18:12:50 walter k3s[19344]: 1b072b5c:  00000000  000462f4 <runtime.fatalthrow+68>  00046154 <runtime.throw+92>  1b072b6c
Dec 18 18:12:50 walter k3s[19344]: 1b072b6c:  000748fc <runtime.fatalthrow.func1+0>  0fa7fc00  00046154 <runtime.throw+92>  1b072b7c
Dec 18 18:12:50 walter k3s[19344]: 1b072b7c:  0005d688 <runtime.sigpanic+784>  1b072b84  00074890 <runtime.throw.func1+0>  03372b14
Dec 18 18:12:50 walter k3s[19344]: 1b072b8c:  00000005  0207c614 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus+88>  03372b14  00000005
Dec 18 18:12:50 walter k3s[19344]: 1b072b9c:  00000001  1225ccd0  0fa7fc00  0207c614 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus+88>
Dec 18 18:12:50 walter k3s[19344]: 1b072bac: <00000000  00003e01  652228d8  00000000
Dec 18 18:12:50 walter k3s[19344]: 1b072bbc:  000000db  1225ccd0  00085684 <sync.(*Pool).pin+48>  000000dc
Dec 18 18:12:50 walter k3s[19344]: 1b072bcc:  00000013  76fab750  00085684 <sync.(*Pool).pin+48>  0008524c <sync.(*Pool).Put+40>
Dec 18 18:12:50 walter k3s[19344]: 1b072bdc:  05577fe4  00000004  00000600  00000002
Dec 18 18:12:50 walter k3s[19344]: 1b072bec:  00117e10 <fmt.(*pp).free+160>  00022e08 <runtime.mapassign_faststr+72>  1b421b00  29831cc6
Dec 18 18:12:50 walter k3s[19344]: 1b072bfc:  00000024  8b46bf36  02079030 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.syncOne+412>  1b072c40
Dec 18 18:12:50 walter k3s[19344]: 1b072c0c:  29831cc6  8b46bf36  0a11b200  0000008b
Dec 18 18:12:50 walter k3s[19344]: 1b072c1c:  1beffb60  00000013  8b46bf36  00000013
Dec 18 18:12:50 walter k3s[19344]: 1b072c2c:  1beffb60  1b0731bc  0207ce60 <github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.(*Controller).syncAll.func4+208>  1b073210
Dec 18 18:12:50 walter k3s[19344]: 1b072c3c:  00000000  00000000  00000000  1bfaaaa0
Dec 18 18:12:50 walter k3s[19344]: 1b072c4c:  0000000b  00000000  00000000  1bfaaa98
Dec 18 18:12:50 walter k3s[19344]: 1b072c5c: >00000007  1bf989c0  0000003b  1bb9ad50
Dec 18 18:12:50 walter k3s[19344]: 1b072c6c:  00000024  1bfaaab0  00000008  00000000
Dec 18 18:12:50 walter k3s[19344]: 1b072c7c:  00000000  00000000  00000000  d744e84f
Dec 18 18:12:50 walter k3s[19344]: 1b072c8c:  0000000e  057d1290  00000000  00000000
Dec 18 18:12:50 walter k3s[19344]: 1b072c9c:  1bfb26e0  1bfb2700  00000000  00000000
Dec 18 18:12:50 walter k3s[19344]: 1b072cac:  00000000  00000000  00000000  00000000
Dec 18 18:12:50 walter k3s[19344]: 1b072cbc:  00000000  00000000  1e00e410  00000002
Dec 18 18:12:50 walter k3s[19344]: 1b072ccc:  00000002  1bfaab70  00000009  00000000
Dec 18 18:12:50 walter k3s[19344]: github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.getFinishedStatus(0x1bf989c0, 0x3b, 0x1bb9ad50, 0x24)
```

**Comment by brandond** [comment]
Do the 2GB models have a different CPU in them or something?
```
sysadm@pi01:~$ tail -n 4 /proc/cpuinfo
Hardware	: BCM2835
Revision	: c03112
Serial		: 10000000498ce2e3
Model		: Raspberry Pi 4 Model B Rev 1.2
```

I have Ubuntu 20.10 on all my Pi's at the moment, I can try putting raspbian on them to see if that behaves differently.

**Comment by dweomer** [comment]
It might be worth wiping the directory pointed to by the "current" symlink under your data-dir (and also the symlink) after stopping but before restarting k3s (it will be recreated).

**Comment by davidnuzik** [comment]
I'm unable to reproduce on my 32-bit (Raspbian) Raspi 4 4GB model

```
pi@davidspi4:~ $ uname -a
Linux davidspi4 5.4.79-v7l+ #1373 SMP Mon Nov 23 13:27:40 GMT 2020 armv7l GNU/Linux

pi@davidspi4:~ $ tail -n4 /proc/cpuinfo
Hardware        : BCM2711
Revision        : c03111
Serial          : 10000000fb7fa440
Model           : Raspberry Pi 4 Model B Rev 1.1
```

**Comment by narqo** [comment]
For my Pi4 it shows the following

```
pi@pi-41:~ $ tail -n 4 /proc/cpuinfo
Hardware	: BCM2711
Revision	: b03111
Serial		: 100000005909b92c
Model		: Raspberry Pi 4 Model B Rev 1.1
```

**Comment by clambin** [comment]
I have the same problem running the manager in a 4GB RPI4 server.

```
$ tail -n 4 /proc/cpuinfo
Hardware	: BCM2711
Revision	: c03112
Serial		: 10000000e86fd369
Model		: Raspberry Pi 4 Model B Rev 1.2
```

**Comment by brandond** [comment]
@clambin what Linux distribution are you running? What is the output of `uname -a` ?

**Comment by fubsle** [comment]
Same issue here. It started after upgrading from 1.19.4 to v1.19.5+k3s2.
A clean k3s instance is running fine...

I can reproduce it by simply creating a cronjob - like [this example from k8s doc ](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#example)

* I am running Raspbian Lite
* Installed k3s with https://github.com/k3s-io/k3s-ansible
```
$ tail -n 4 /proc/cpuinfo
Hardware	: BCM2711
Revision	: c03112
Serial		: 10000000cfaedbe0
Model		: Raspberry Pi 4 Model B Rev 1.2
```
```
$ uname -a
Linux k3s-0 5.4.72-v7l+ #1356 SMP Thu Oct 22 13:57:51 BST 2020 armv7l GNU/Linux
```

Here's the syslog from the point when cronjob object was created until k3s service exits 
[syslog.gz](https://github.com/k3s-io/k3s/files/5751994/syslog.gz)

**Comment by clambin** [comment]
> @clambin what Linux distribution are you running? What is the output of `uname -a` ?

Standard raspbian:
 
```
$ uname -a
Linux raspberrypi1 5.4.79-v7l+ #1373 SMP Mon Nov 23 13:27:40 GMT 2020 armv7l GNU/Linux
```

**Comment by brandond** [comment]
If possible, could someone experiencing this issue try a dev build?
```
curl -sfL https://get.k3s.io | INSTALL_K3S_COMMIT=2d1a60a9dd450b3ef4c6a9b6b813319d51175e65 bash -
```

**Comment by clambin** [comment]
@brandond downloaded the binary directly and running it now.  got a bunch of warnings/errors on startup that I'm going through, but so far haven't seen any panics.  will leave it running for a while and let you know.

**Comment by clambin** [comment]
No issues after a day of running.

**Comment by brandond** [comment]
Interesting. That commit has a work-around for https://github.com/kubernetes/kubernetes/issues/97685 which appears to be a memory corruption issue on 32bit platforms. This will most likely need to be fixed in golang itself, but it is interesting that this also fixes the unrelated crashes in other parts of the Kubernetes runtime.

**Comment by narqo** [comment]
@bradtopol to make sure I look in the right place 2d1a60a9dd450b3ef4c6a9b6b813319d51175e65 is a patch version of k8s 1.20, right? Above I mentioned that I hadn't seen the issue with `v1.20.0-rc3+k3s.1`. 

I also tried 1.20.0+k3s2, but I've seen some weird behaviour, possible explained in kubernetes/kubernetes#97685 so I've decided to roll back but at least the server didn't crash.

**Comment by narqo** [comment]
I've tried v1.19.7-rc1+k3s1 on my cluster and got the same panic but this time, it seems upstream k8s added a panic handler, which prints the trace log in a more clean way:

```
Jan 14 19:58:29 pi-41 k3s[24668]: E0114 19:58:29.687218   24668 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer d
Jan 14 19:58:29 pi-41 k3s[24668]: goroutine 15969 [running]:
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x2f8d0d8, 0x5578c40)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x74
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x6c
Jan 14 19:58:29 pi-41 k3s[24668]: panic(0x2f8d0d8, 0x5578c40)
Jan 14 19:58:29 pi-41 k3s[24668]:         /usr/local/go/src/runtime/panic.go:969 +0x158
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.syncOne(0x16a9d500, 0x163b0000, 0x4, 0x4, 0x68ecaf50, 0xbff842f5, 0xf6b1b0c6, 0x8, 0x57d72c8, 0x3a6
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob/cronjob_controller.go:216 +0x34
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.(*Controller).syncAll.func4(0x3a33008, 0x16a9d500, 0x16a9d500, 0x3a33008)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob/cronjob_controller.go:140 +0xd0
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/api/meta.EachListItem(0x3a33020, 0x17817000, 0x158e1e50, 0x7bfee01, 0xb90101)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/api/meta/help.go:151 +0x3a8
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/client-go/tools/pager.(*ListPager).EachListItem.func1(0x3a33020, 0x17817000, 0x17816f01, 0x17816f80)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/client-go/tools/pager/pager.go:158 +0x2c
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/client-go/tools/pager.(*ListPager).eachListChunkBuffered(0x163ac680, 0x3a5cb18, 0x163ac6a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/client-go/tools/pager/pager.go:203 +0x154
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/client-go/tools/pager.(*ListPager).EachListItem(0x163ac680, 0x3a5cb38, 0x7456028, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/client-go/tools/pager/pager.go:157 +0x60
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.(*Controller).syncAll(0xc4dd4d0)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob/cronjob_controller.go:135 +0x3d8
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xfc54628)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x54
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xfc54628, 0x3a27bd8, 0x13c17320, 0xd0c4a01, 0x0)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x74
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xfc54628, 0x540be400, 0x2, 0x0, 0x0, 0x1, 0x0)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x80
Jan 14 19:58:29 pi-41 k3s[24668]: github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xfc54628, 0x540be400, 0x2, 0x0)
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0x48
Jan 14 19:58:29 pi-41 k3s[24668]: created by github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob.(*Controller).Run
Jan 14 19:58:29 pi-41 k3s[24668]:         /go/src/github.com/rancher/k3s/vendor/k8s.io/kubernetes/pkg/controller/cronjob/cronjob_controller.go:98 +0xf8
```

Note, similar to as before, the panic starts in `/go/···/k8s.io/kubernetes/pkg/controller/cronjob/cronjob_controller.go:216`, which still feels to me, that server _somehow_ receives a `nil` object instead of `*batchv1beta1.CronJob`. Could it be that the etcd-shim (I use the default sqlite3 via kine) somehow returns a broken record when k8s iterates over the results of its `cronJobListFunc`? Anything I can check to help debugging that, @brandond?

**Comment by brandond** [comment]
See if this is fixed on v1.20.2+k3s1. The golang compiler bug is still out there and Kubernetes probably won't update to a release that has it fixed for another couple months yet. If you're still seeing crashes in the CronJob controller then it might be triggering the same thing and need a temporary fix to avoid it.

**Comment by erikwilson** [comment]
Compiling & testing with optimizations disabled (build with `-gcflags '-N'`) might be helpful until the golang fix is released.

**Comment by narqo** [comment]
> See if this is fixed on v1.20.2+k3s1

Running `v1.20.2+k3s1` on the server node for that last 24h, and it seems to work fine 🎉

**Comment by mccare** [comment]
I've seen SIGILL and k3s server dying when adding a cronjob with version `v1.19.7+k3s1` (as described above). I've upgraded to `v1.20.2+k3s1` and the problem went away. RPi 4 with 8GB on raspbian (Linux nsa 5.4.83-v7l+ #1379 SMP Mon Dec 14 13:11:54 GMT 2020 armv7l GNU/Linux)

**Comment by dereknola** [comment]
Closing this is it has been fixed in the v1.20 release chain  and v1.19 is EOL
